content/blog/2022-04-07-compiling-black-with-mypyc-part-2.md

```diff
@@ -0,0 +1,537 @@
+-from blib2to3.pytree import NL, Context, RawNode, Leaf, Node
++from blib2to3.pytree import convert, NL, Context, RawNode, Leaf, Node
+
+@@ -199,16 +206,13 @@ class Parser(object):
+             raise ParseError("bad token", type, value, context)
+         return ilabel
+
+     def shift(self, type: int, value: Text, newstate: int, context: Context) -> None:
+         """Shift a token.  (Internal)"""
+         dfa, state, node = self.stack[-1]
+         rawnode: RawNode = (type, value, context, None)
+-        newnode = self.convert(self.grammar, rawnode)
+-        if newnode is not None:
+-            assert node[-1] is not None
+-            node[-1].append(newnode)
++        newnode = convert(self.grammar, rawnode)
++        assert node[-1] is not None
++        node[-1].append(newnode)
+         self.stack[-1] = (dfa, newstate, node)
+
+     def push(self, type: int, newdfa: DFAS, newstate: int, context: Context) -> None:
+     @@ -221,12 +225,11 @@ class Parser(object):
+     def pop(self) -> None:
+         """Pop a nonterminal.  (Internal)"""
+         popdfa, popstate, popnode = self.stack.pop()
+-        newnode = self.convert(self.grammar, popnode)
+-        if newnode is not None:
+-            if self.stack:
+-                dfa, state, node = self.stack[-1]
+-                assert node[-1] is not None
+-                node[-1].append(newnode)
+-            else:
+-                self.rootnode = newnode
+-                self.rootnode.used_names = self.used_names
++        newnode = convert(self.grammar, popnode)
++        if self.stack:
++            dfa, state, node = self.stack[-1]
++            assert node[-1] is not None
++            node[-1].append(newnode)
++        else:
++            self.rootnode = newnode
++            self.rootnode.used_names = self.used_names
+```
+
+Also since `blib2to3.pytree.convert` is guaranteed to return a non-None value, I could
+drop a branch which was neat :)
+
+______________________________________________________________________
+
+### Detour: let's build developer tooling!
+
+If you've worked with me before you'll know that I *love* to write developer tooling to
+make life easier. (Un)fortunately this project needed two bits of tooling that simply
+didn't exist at the time:
+
+1. A benchmark suite
+1. A tool to compare two builds of Black behaviourally
+
+The first one is pretty self-explanatory, I needed a good benchmark suite to make sure
+this project would actually improve performance, and to also quantify the gains. The
+latter was important when weighing optimizations.
+
+The second one is a less clear, effectively I wanted [mypy-primer], but for Black:
+
+![mypy-primer comment on python/mypy PR 12064 describing the impact of the change](/media/mypy-primer-comment.png)
+
+> [mypy-primer comment on mypy PR #12064][mypy-primer-comment].
+
+So I got to work creating [blackbench] and
+[(the original) diff-shades][original-diff-shades]. Honestly, in hindsight blackbench
```

Comment:
```suggestion
[(the original) diff-shades][original-diff-shades]. Honestly, in hindsight, blackbench
```
I would also suggest removing the "Honestly" too.

Reply:
```suggestion
[(the original) diff-shades][original-diff-shades]. In hindsight, blackbench
```